From 87076e82e26688f43d3b64939ba2b6034743f59e Mon Sep 17 00:00:00 2001
From: Bo Shen <voice.shen@atmel.com>
Date: Wed, 29 Oct 2014 15:14:41 +0800
Subject: [PATCH] ARM: atmel: sama5d4: move get_cpu_name to device file

As the SAMA5D4 family has SAMA5D41, SAMA5D42, SAMA5D43 and SAMA5D44
members, we need to distinguish them. So, using get_cpu_name to get
the real name of them.

Signed-off-by: Bo Shen <voice.shen@atmel.com>
---
 arch/arm/cpu/armv7/at91/Makefile                |  1 +
 arch/arm/cpu/armv7/at91/sama5d4_devices.c       | 31 +++++++++++++++++++++++++
 arch/arm/include/asm/arch-at91/sama5d4.h        | 17 ++++++++++++++
 board/atmel/sama5d4_xplained/sama5d4_xplained.c |  5 ----
 board/atmel/sama5d4ek/sama5d4ek.c               |  5 ----
 5 files changed, 49 insertions(+), 10 deletions(-)
 create mode 100644 arch/arm/cpu/armv7/at91/sama5d4_devices.c

diff --git a/arch/arm/cpu/armv7/at91/Makefile b/arch/arm/cpu/armv7/at91/Makefile
index 0a2e48d..f4f35a4 100644
--- a/arch/arm/cpu/armv7/at91/Makefile
+++ b/arch/arm/cpu/armv7/at91/Makefile
@@ -9,6 +9,7 @@
 #
 
 obj-$(CONFIG_SAMA5D3)	+= sama5d3_devices.o
+obj-$(CONFIG_SAMA5D4)	+= sama5d4_devices.o
 obj-y += clock.o
 obj-y += cpu.o
 obj-y += reset.o
diff --git a/arch/arm/cpu/armv7/at91/sama5d4_devices.c b/arch/arm/cpu/armv7/at91/sama5d4_devices.c
new file mode 100644
index 0000000..4873b50
--- /dev/null
+++ b/arch/arm/cpu/armv7/at91/sama5d4_devices.c
@@ -0,0 +1,31 @@
+/*
+ * Copyright (C) 2014 Atmel
+ *		      Bo Shen <voice.shen@atmel.com>
+ *
+ * SPDX-License-Identifier:	GPL-2.0+
+ */
+
+#include <common.h>
+#include <asm/io.h>
+#include <asm/arch/sama5d4.h>
+
+char *get_cpu_name()
+{
+	unsigned int extension_id = get_extension_chip_id();
+
+	if (cpu_is_sama5d4())
+		switch (extension_id) {
+		case ARCH_EXID_SAMA5D41:
+			return "SAMA5D41";
+		case ARCH_EXID_SAMA5D42:
+			return "SAMA5D42";
+		case ARCH_EXID_SAMA5D43:
+			return "SAMA5D43";
+		case ARCH_EXID_SAMA5D44:
+			return "SAMA5D44";
+		default:
+			return "Unknown CPU type";
+		}
+	else
+		return "Unknown CPU type";
+}
diff --git a/arch/arm/include/asm/arch-at91/sama5d4.h b/arch/arm/include/asm/arch-at91/sama5d4.h
index 38fade5..dc4d0dd 100644
--- a/arch/arm/include/asm/arch-at91/sama5d4.h
+++ b/arch/arm/include/asm/arch-at91/sama5d4.h
@@ -174,6 +174,23 @@
 #define CPU_HAS_PCR
 #define CPU_HAS_H32MXDIV
 
+/* sama5d4 series chip id definitions */
+#define ARCH_ID_SAMA5D4		0x8a5c07c0
+#define ARCH_EXID_SAMA5D41	0x00000001
+#define ARCH_EXID_SAMA5D42	0x00000002
+#define ARCH_EXID_SAMA5D43	0x00000003
+#define ARCH_EXID_SAMA5D44	0x00000004
+
+#define cpu_is_sama5d4()	(get_chip_id() == ARCH_ID_SAMA5D4)
+#define cpu_is_sama5d41()	(cpu_is_sama5d4() && \
+		(get_extension_chip_id() == ARCH_EXID_SAMA5D41))
+#define cpu_is_sama5d42()	(cpu_is_sama5d4() && \
+		(get_extension_chip_id() == ARCH_EXID_SAMA5D42))
+#define cpu_is_sama5d43()	(cpu_is_sama5d4() && \
+		(get_extension_chip_id() == ARCH_EXID_SAMA5D43))
+#define cpu_is_sama5d44()	(cpu_is_sama5d4() && \
+		(get_extension_chip_id() == ARCH_EXID_SAMA5D44))
+
 /*
  * The clock of peripheral connect to H32MX is decided by H32MXDIV
  * bit in MCKR, if this bit is set, the peripheral only work at half
diff --git a/board/atmel/sama5d4_xplained/sama5d4_xplained.c b/board/atmel/sama5d4_xplained/sama5d4_xplained.c
index d7fb9bc9..de0baad 100644
--- a/board/atmel/sama5d4_xplained/sama5d4_xplained.c
+++ b/board/atmel/sama5d4_xplained/sama5d4_xplained.c
@@ -251,11 +251,6 @@ void sama5d4_xplained_macb0_hw_init(void)
 }
 #endif
 
-char *get_cpu_name()
-{
-	return "SAMA5D4";
-}
-
 static void sama5d4_xplained_serial3_hw_init(void)
 {
 	at91_set_b_periph(AT91_PIO_PORTE, 17, 1);	/* TXD3 */
diff --git a/board/atmel/sama5d4ek/sama5d4ek.c b/board/atmel/sama5d4ek/sama5d4ek.c
index dc12e04..daed5cb 100644
--- a/board/atmel/sama5d4ek/sama5d4ek.c
+++ b/board/atmel/sama5d4ek/sama5d4ek.c
@@ -269,11 +269,6 @@ void sama5d4ek_macb1_hw_init(void)
 }
 #endif
 
-char *get_cpu_name()
-{
-	return "SAMA5D4";
-}
-
 static void sama5d4ek_serial3_hw_init(void)
 {
 	at91_set_b_periph(AT91_PIO_PORTE, 17, 1);	/* TXD3 */
-- 
1.9.1

